java-util/src/main/java/io/druid/java/util/common/granularity/Granularity.java

```diff
@@ -0,0 +1,538 @@
+    MONTH(
+        "'dt'=yyyy-MM",
+        "'y'=yyyy/'m'=MM",
+        "'y'=yyyy/'m'=MM"
+    ),
+    QUARTER(MONTH),
+    YEAR(
+        "'dt'=yyyy",
+        "'y'=yyyy",
+        "'y'=yyyy"
+    ),
+    ALL(),
+    NONE();
+
+    private final String hiveFormat;
+    private final String lowerDefaultFormat;
+    private final String defaultFormat;
+
+    GranularityType()
+    {
+      this(null, null, null);
+    }
+
+    GranularityType(
+        final String hiveFormat,
+        final String lowerDefaultFormat, final String defaultFormat
+    )
+    {
+      this.hiveFormat = hiveFormat;
+      this.lowerDefaultFormat = lowerDefaultFormat;
+      this.defaultFormat = defaultFormat;
+    }
+
+    GranularityType(GranularityType granularityType)
+    {
+      this(
+          granularityType.getHiveFormat(),
+          granularityType.getLowerDefaultFormat(),
+          granularityType.getDefaultFormat()
+      );
+    }
+
+    static Granularity createGranularity(String str)
+    {
+      return createGranularity(str, null, null);
+    }
+
+    static Granularity createGranularity(String str, DateTime origin, DateTimeZone tz)
+    {
+      GranularityType granularityType = GranularityType.valueOf(str);
+
+      switch (granularityType) {
+        case SECOND:
+          return new PeriodGranularity(new Period("PT1S"), origin, tz);
+        case MINUTE:
+          return new PeriodGranularity(new Period("PT1M"), origin, tz);
+        case FIVE_MINUTE:
+          return new PeriodGranularity(new Period("PT5M"), origin, tz);
+        case TEN_MINUTE:
+          return new PeriodGranularity(new Period("PT10M"), origin, tz);
+        case FIFTEEN_MINUTE:
+          return new PeriodGranularity(new Period("PT15M"), origin, tz);
+        case THIRTY_MINUTE:
+          return new PeriodGranularity(new Period("PT30M"), origin, tz);
+        case HOUR:
+          return new PeriodGranularity(new Period("PT1H"), origin, tz);
+        case SIX_HOUR:
+          return new PeriodGranularity(new Period("PT6H"), origin, tz);
+        case DAY:
+          return new PeriodGranularity(new Period("P1D"), origin, tz);
+        case WEEK:
+          return new PeriodGranularity(new Period("P1W"), origin, tz);
+        case MONTH:
+          return new PeriodGranularity(new Period("P1M"), origin, tz);
+        case QUARTER:
+          return new PeriodGranularity(new Period("P3M"), origin, tz);
+        case YEAR:
+          return new PeriodGranularity(new Period("P1Y"), origin, tz);
+        case ALL:
+          return new AllGranularity();
+        case NONE:
+          return new NoneGranularity();
+        default:
+          throw new IAE("[%s] granularity not supported with strings. Try with Period instead", str);
+      }
+    }
+
+    // Note: This is only an estimate based on the values in period.
+    // This will not work for complicated periods that represent say 1 year 1 day
+    public static GranularityType fromPeriod(Period period)
```

Comment:
Why is this called?  I'm wondering if the throwing of `IAE` if it doesn't work is going to surprise call sites instead of doing something like returning `null` to indicate that it wasn't able to find one.